nncf/common/tensor_new/tensor.py

```diff
@@ -0,0 +1,190 @@
+    from nncf.torch import torch_ops
+
+except ImportError:
+    torch_ops = None
+
+
+DataType = TypeVar("DataType")
+DeviceType = TypeVar("DeviceType")
+
+
+FUNC_MAP_DISPATCHER = {
+    TensorBackendType.NUMPY: numpy_ops,
+}
+if torch_ops:
+    FUNC_MAP_DISPATCHER[TensorBackendType.TORCH] = torch_ops
+
+
+class Tensor:
+    """
+    An interface to framework specific tensors for common NNCF algorithms.
+    """
+
+    def __init__(self, data: Optional[DataType]):
+        self._data = data.data if isinstance(data, Tensor) else data
+
+    @property
+    def data(self) -> DataType:
+        return self._data
+
+    @property
+    def shape(self) -> List[int]:
+        return Tensor(list(self.data.shape))
+
+    def __bool__(self) -> bool:
+        return bool(self.data)
+
+    def __iter__(self) -> Iterator:
+        return iter(self.data)
+
+    def __getitem__(self, index: int) -> "Tensor":
+        return Tensor(self.data[index])
+
+    # built-in operations
+
+    def __add__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data + unwrap_tensor_data(other))
+
+    def __radd__(self, other: DataType) -> "Tensor":
+        return Tensor(unwrap_tensor_data(other) + self.data)
+
+    def __sub__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data - unwrap_tensor_data(other))
+
+    def __rsub__(self, other: DataType) -> "Tensor":
+        return Tensor(unwrap_tensor_data(other) - self.data)
+
+    def __mul__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data * unwrap_tensor_data(other))
+
+    def __rmul__(self, other: DataType) -> "Tensor":
+        return Tensor(unwrap_tensor_data(other) * self.data)
+
+    def __pow__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data ** unwrap_tensor_data(other))
+
+    def __truediv__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data / unwrap_tensor_data(other))
+
+    def __floordiv__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data // unwrap_tensor_data(other))
+
+    def __neg__(self) -> "Tensor":
+        return Tensor(-self.data)
+
+    # Comparison operators
+
+    def __lt__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data < unwrap_tensor_data(other))
+
+    def __le__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data <= unwrap_tensor_data(other))
+
+    def __eq__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data == unwrap_tensor_data(other))
+
+    def __nq__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data != unwrap_tensor_data(other))
+
+    def __gt__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data > unwrap_tensor_data(other))
+
+    def __ge__(self, other: DataType) -> "Tensor":
+        return Tensor(self.data >= unwrap_tensor_data(other))
+
+    # Tensor functions
+
+    @property
+    def device(self) -> Optional[DeviceType]:
+        return tensor_func_dispatcher("device", self.data)
```

Comment:
The `DeviceType` should be `Enum` if `device` property will be used in common code.
```python
if tensor.device == DeviceType.CPU:
    # do something
```
Otherwise, it is not clear how to work with backend-specific `DeviceType`.

By the way, could you please give more details about why we should wrap return value (for this method) into `Tensor` object inside the `tensor_func_dispatcher()` methond? Is it bug?

Reply:
done